Rsite/source/tutorial/CustomLossGluon.md

```diff
@@ -0,0 +1,235 @@
+
+training_dir = "images_background/Tifinagh"
+testing_dir = "images_background/Inuktitut_(Canadian_Aboriginal_Syllabics)"
+train_dataset = GetImagePairs(training_dir)
+test_dataset = GetImagePairs(testing_dir)
+train_dataloader = gluon.data.DataLoader(train_dataset.transform(transform), shuffle=True, num_workers=2, batch_size=16)
+test_dataloader = gluon.data.DataLoader(test_dataset.transform(transform), shuffle=True, num_workers=2, batch_size=1)
+```
+
+Following code plots some examples from the test dataset. 
+
+
+```python
+img1, img2, label = test_dataset[0]
+print("Same: {}".format(int(label.asscalar()) == 0))
+fig, (ax0, ax1) = plt.subplots(ncols=2, figsize=(10, 5)) 
+ax0.imshow(img1.asnumpy()[:,:,0], cmap='gray')
+ax0.axis('off')
+ax1.imshow(img2.asnumpy()[:,:,0], cmap='gray')
+ax1.axis("off")
+plt.show()
+
+```
+
+    Same: False
+
+
+
+![png](CustomLossGluon1.png)
+
+
+### Train the Siamese network
+
+Before we can start training, we need to instatiate the custom constructive loss function and initialize the model.
+
+
+```python
+model = Siamese()
+model.initialize(init=mx.init.Xavier())
+trainer = gluon.Trainer(model.collect_params(), 'adam', {'learning_rate': 0.001})
+loss = ContrastiveLoss(margin=6.0)
+```
+
+Start the training loop:
+
+
+```python
+for epoch in range(10):
+    
+    for i, data in enumerate(train_dataloader):
+        
+        image1, image2, label = data
+        with autograd.record():
+            output1, output2 = model(image1, image2)
+            loss_contrastive = loss(output1, output2, label)
+        loss_contrastive.backward()
+        trainer.step(image1.shape[0])
+        
+        print("Epoch number {}\n Current loss {}\n".format(epoch, loss_contrastive.mean().asscalar()))
+```
+
+### Test the trained Siamese network
+During inference we compute the Euclidean distance between the output vectors of the Siamese network. High distance indicates dissimilarity, low values indicate similarity.  
+
+
+```python
+for i, data in enumerate(test_dataloader):
+
+    img1, img2, label = data
+    output1, output2 = model(img1, img2)
+
+    print("Euclidean Distance:", mx.ndarray.sqrt(mx.ndarray.sum(mx.ndarray.square(output1 - output2))).asscalar(), "Test label", label[0].asscalar())
+    
+    fig, (ax0, ax1) = plt.subplots(ncols=2, figsize=(10, 5)) 
+    ax0.imshow(img1.asnumpy()[0,0,:,:], cmap='gray')
+    ax0.axis('off')
+    ax1.imshow(img2.asnumpy()[0,0,:,:], cmap='gray')
+    ax1.axis("off")
+    plt.show()
+
+    break
+```
+
+    Euclidean Distance: 2.494767 Test label 1.0
+
+
+
+![png](CustomLossGluon2.png)
+
+
+### Common Pitfalls with custom loss functions
+
+When customizing loss functions, we may encounter certain pitfalls. If the loss is not decreasing as expected or if forward/backward pass is crashing, then one should check the following:
+
+#### Activation function in the last layer:
```

Comment:
some of the section headers end in colons and some don't.
Also the capitalization of Common Pitfalls with custom loss functions doesn't match the capitalization of other headers